samza-core/src/main/java/org/apache/samza/operators/impl/WindowOperatorImpl.java

```diff
@@ -18,26 +18,323 @@
+    long now = System.currentTimeMillis();
+    TriggerTimerState state;
+    while ((state = pendingCallbacks.peek()) != null && state.getScheduleTimeMs() < now) {
+      pendingCallbacks.remove();
+      state.getCallback().run();
+
+      TriggerImplWrapper wrapper = triggers.get(state.triggerKey);
+      if (wrapper != null) {
+        wrapper.onTimer(collector, coordinator);
+      }
+    }
+  }
+
+  private WindowKey<K> getStoreKey(M message) {
+    Function<M, K> keyExtractor = window.getKeyExtractor();
+    K key = null;
+
+    if (keyExtractor != null) {
+      key = keyExtractor.apply(message);
+    }
+
+    String windowId = null;
+    if (window.getWindowType() == WindowType.TUMBLING) {
+      long triggerDurationMs = ((TimeTrigger<M>) window.getDefaultTrigger()).getDuration().toMillis();
+      final long now = System.currentTimeMillis();
+      Long windowBoundary = now - now % triggerDurationMs;
+      windowId = windowBoundary.toString();
+    }
+
+    return new WindowKey<>(key, windowId);
+  }
+
+  private TriggerImplWrapper getOrCreateTriggerWrapper(TriggerKey<K> triggerKey, Trigger<M> trigger) {
+    TriggerImplWrapper wrapper = triggers.get(triggerKey);
+    if (wrapper != null) {
+      return wrapper;
+    }
+
+    TriggerHandlerImpl triggerHandler = new TriggerHandlerImpl();
+    TriggerImpl<M> triggerImpl = TriggerImpls.createTriggerImpl(trigger);
+    TriggerContextImpl triggerContext = new TriggerContextImpl(triggerKey);
+    wrapper = new TriggerImplWrapper(triggerKey, triggerImpl, triggerContext, triggerHandler);
+    triggers.put(triggerKey, wrapper);
+
+    return wrapper;
+  }
+
+  private void handleTrigger(TriggerKey<K> triggerKey, MessageCollector collector, TaskCoordinator coordinator) {
+    TriggerImplWrapper wrapper = triggers.get(triggerKey);
+
+    // Cancel all early triggers too when the default trigger fires.
+    if (triggerKey.getType() == TriggerType.DEFAULT) {
+      cancelTrigger(triggerKey);
+      cancelTrigger(new TriggerKey(TriggerType.EARLY, triggerKey.getKey()));
+    }
+
+    // Cancel non-repeating early triggers.
+    if (triggerKey.getType() == TriggerType.EARLY && !wrapper.isRepeating()) {
+      cancelTrigger(triggerKey);
+    }
+
+    WindowKey<K> windowKey = triggerKey.key;
+    WV wv = store.get(windowKey);
+
+    if (wv == null) {
+      return;
+    }
+
+    WindowPane<K, WV> paneOutput = new WindowPane<>(windowKey, wv, window.getAccumulationMode());
+    // Handle accumulation modes.
+    if (window.getAccumulationMode() == AccumulationMode.DISCARDING) {
+      store.put(windowKey, null);
+    }
+
+    if (paneOutput.getMessage() instanceof Collection) {
+      WV valCopy = (WV) new ArrayList<>((Collection<WV>) paneOutput.getMessage());
+      paneOutput = new WindowPane(windowKey, valCopy, window.getAccumulationMode());
+    }
+
+    WindowOperatorImpl.super.propagateResult((WM) paneOutput, collector, coordinator);
```

Comment:
Good point. It was previously done directly in the handler's `onTrigger` function. So, it required the `super` object to be accessed by closure (instead of directly). Now that we don't propagate results in the callback, we can directly use `super`.